flink/src/main/java/org/apache/iceberg/flink/source/StreamingMonitorFunction.java

```diff
@@ -0,0 +1,165 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.iceberg.flink.source;
+
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import org.apache.flink.api.common.state.ListState;
+import org.apache.flink.api.common.state.ListStateDescriptor;
+import org.apache.flink.api.common.typeutils.base.LongSerializer;
+import org.apache.flink.runtime.state.FunctionInitializationContext;
+import org.apache.flink.runtime.state.FunctionSnapshotContext;
+import org.apache.flink.streaming.api.checkpoint.CheckpointedFunction;
+import org.apache.flink.streaming.api.functions.source.RichSourceFunction;
+import org.apache.iceberg.Snapshot;
+import org.apache.iceberg.Table;
+import org.apache.iceberg.flink.TableLoader;
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.util.SnapshotUtil;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * This is the single (non-parallel) monitoring task which takes a {@link FlinkInputFormat},
+ * it is responsible for:
+ *
+ * <ol>
+ *     <li>Monitoring snapshots of the Iceberg table.</li>
+ *     <li>Creating the {@link FlinkInputSplit splits} corresponding to the incremental files</li>
+ *     <li>Assigning them to downstream tasks for further processing.</li>
+ * </ol>
+ *
+ * <p>The splits to be read are forwarded to the downstream {@link StreamingReaderOperator}
+ * which can have parallelism greater than one.
+ */
+public class StreamingMonitorFunction extends RichSourceFunction<FlinkInputSplit> implements CheckpointedFunction {
+
+  private static final Logger LOG = LoggerFactory.getLogger(StreamingMonitorFunction.class);
+
+  private static final long DUMMY_START_SNAPSHOT_ID = -1;
+
+  private final TableLoader tableLoader;
+  private final ScanContext scanContext;
+
+  private volatile boolean isRunning = true;
+  private transient Object checkpointLock;
+  private transient Table table;
+  private transient ListState<Long> snapshotIdState;
+  private transient long startSnapshotId;
```

Comment:
This appears to be used as the last processed snapshot ID. I think it would make more sense to name it `lastSnapshotId` so it is clear that this snapshot has already been processed.